### DISCUSSION TITLE
GET_LOCK fails on Planetscale

### DISCUSSION THREAD

**Discussion by drkNsubuga** [planetscale/discussion#472] [Q&A] [answered]
I connected planetscale with codeigniter which basically runs the following query in the session library.

```php
SELECT GET_LOCK($arg, 300) AS ci_session_lock; //$arg is an md5 string
```
The query specifically fails with Planetscale but succeeds with local mysql. The error log reveals that the result does not contain the `ci_session_lock` column;
When I run the same query in a db client, I get this error:
```console
Error Code: 1105. get_lock(:vtg1, :vtg2) allowed only with dual
```
The error persists even though the query is adjusted to:
 ```php
 SELECT GET_LOCK($arg, 300) AS ci_session_lock FROM dual;
```
What could be the issue?

**Comment by orware** [comment, marked as answer]
Hello @drkNsubuga,

I believe we connected on this topic directly in your ticket earlier this morning.

For others coming to this page in the future, this is related to other similar discussions such as https://github.com/planetscale/discussion/discussions/307 and it has been reported to our Vitess team some months ago so it is currently on the backlog to be eventually tackled by the team there.

**Comment by drkNsubuga** [comment]
Hi @orware,

Thank you so much for the feedback. I understand the issue is with Vitess. I made a quick workaround but 'll consider using other alternatives for session management as well.

**Comment by vsy** [comment]
Dear @drkNsubuga , I'm running into the same issue. Could you perhaps share your workaround used? Thanks for your time!

**Reply by orware** [reply]
@vsy,

I'm not aware of the details of @drkNsubuga's workaround, but for PHP in particular, if you can run the query in such a way that it doesn't rely on the alias name for the return value that is probably the best workaround I can think of at the moment since the query does return a result, but it just doesn't respect the column alias properly which is the issue on the Vitess side on our end.

If this is the [relevant code line](https://github.com/bcit-ci/CodeIgniter/blob/63d037565dd782795021dcbd3b1ca6f8c8a509e4/system/libraries/Session/drivers/Session_database_driver.php#L409):

```php
if ($this->_db->query("SELECT GET_LOCK('".$arg."', 300) AS ci_session_lock")->row()->ci_session_lock)
```

The pulling over the `row()` value and referencing the `ci_session_lock` value at the end is the problematic part since that attribute won't exist as expected in the result for our database at this time.

I am thinking though that you could potentially tweak that bit of code to not rely on the `ci_session_lock` alias name though and use one of the available array options for CodeIgniter instead:
https://codeigniter.com/userguide3/database/results.html#result-arrays

Similarly, I think the line related to[ releasing the lock](https://github.com/bcit-ci/CodeIgniter/blob/63d037565dd782795021dcbd3b1ca6f8c8a509e4/system/libraries/Session/drivers/Session_database_driver.php#L450) would also need to be tweaked too.

**Reply by drkNsubuga** [reply]
Hi @vsy,

This is the quick workaround I applied as suggested by @orware's explanation: 

> If this is the [relevant code line](https://github.com/bcit-ci/CodeIgniter/blob/63d037565dd782795021dcbd3b1ca6f8c8a509e4/system/libraries/Session/drivers/Session_database_driver.php#L409)

Yes, that's line in CI3


> I am thinking though that you could potentially tweak that bit of code to not rely on the ci_session_lock alias name though and use one of the available array options for CodeIgniter instead:
https://codeigniter.com/userguide3/database/results.html#result-arrays

In ```_get_lock``` function, we use the result array to fetch the ci_session_lock alias column name dynamically
```diff
if ($this->_platform === 'mysql')
{
	$arg = md5($session_id.($this->_config['match_ip'] ? '_'.$_SERVER['REMOTE_ADDR'] : ''));
-	if ($this->_db->query("SELECT GET_LOCK('".$arg."', 300) AS ci_session_lock")->row()->ci_session_lock)
-	{
+		$query = $this->_db->query("SELECT GET_LOCK(?, 300) AS ci_session_lock FROM dual", array($arg));
+
+		// Retrieve the column name dynamically
+       $result = $query->row_array();
+       $column_name = array_key_first($result);
+       $ci_session_lock = $result[$column_name];

+		if ($ci_session_lock) {
			$this->_lock = $arg;
				return TRUE;
		}
```

> Similarly, I think the line related to releasing the lock would also need to be tweaked too.

Yes, the same change is made in the ``release_lock`` function

**Reply by vsy** [reply]
Hi @orware @drkNsubuga , thank you for the swift responses. As the issue at hand doesn't directly involve Planetscale but modifying CodeIgniter, I really appreciate your input and effort in this topic.

Circumventing the alias did the trick indeed! I ended up making the following changes to the Database Session Driver for respectively getting and releasing the lock:

```
if ($this->_platform === 'mysql')
{
    $arg = md5($session_id.($this->_config['match_ip'] ? '_'.$_SERVER['REMOTE_ADDR'] : ''));
    $query = "SELECT GET_LOCK('".$arg."', 300)";
    $result = $this->_db->query($query);
    $row = $result->row_array(); // Get the result as an associative array
    $values = array_values($row); // Get the values of the array
    if ($values[0]) // Access the first value, which should be the result of GET_LOCK
    {
        $this->_lock = $arg;
        return TRUE;
    }

    return FALSE;
}
```

```
if ($this->_platform === 'mysql')
{
    $query = "SELECT RELEASE_LOCK('".$this->_lock."')";
    $result = $this->_db->query($query);
    $row = $result->row_array(); // Get the result as an associative array
    $values = array_values($row); // Get the values of the array
    if ($values[0]) // Access the first value, which should be the result of RELEASE_LOCK
    {
        $this->_lock = FALSE;
        return TRUE;
    }

    return FALSE;
}
```
